whois-update/src/main/java/net/ripe/db/whois/update/generator/SourceGenerator.java

```diff
@@ -21,30 +28,69 @@ public class SourceGenerator extends AttributeGenerator {
 
     private final AuthoritativeResourceData authoritativeResourceData;
 
+    private final RpslObjectDao rpslObjectDao;
     private final CIString source;
     private final CIString nonAuthSource;
 
     @Autowired
     public SourceGenerator(final AuthoritativeResourceData authoritativeResourceData,
                            @Value("${whois.source}") final String source,
-                           @Value("${whois.nonauth.source}") final String nonAuthSource) {
+                           @Value("${whois.nonauth.source}") final String nonAuthSource,
+                           final RpslObjectDao rpslObjectDao) {
         this.authoritativeResourceData = authoritativeResourceData;
         this.source = ciString(source);
         this.nonAuthSource = ciString(nonAuthSource);
+        this.rpslObjectDao = rpslObjectDao;
     }
 
     @Override
     public RpslObject generateAttributes(RpslObject originalObject, RpslObject updatedObject, Update update, UpdateContext updateContext) {
-        switch (updatedObject.getType()) {
-            case AUT_NUM:
-            case ROUTE:
-            case ROUTE6:
-                return generateSource(updatedObject, update, updateContext);
-            default:
+        return switch (updatedObject.getType()) {
+            case AS_SET -> generateAsSetSource(updatedObject, update, updateContext);
+            case AUT_NUM, ROUTE, ROUTE6 -> generateSource(updatedObject, update, updateContext);
+            default -> updatedObject;
+        };
+    }
+
+    private RpslObject generateAsSetSource(final RpslObject updatedObject, final Update update, final UpdateContext updateContext) {
+        if (update.getOperation() == Operation.DELETE) {
+            return updatedObject;
+        }
+        final String asSetKey = updatedObject.getKey().toString();
+
+        final boolean flatAsSet = !asSetKey.contains(":");
+        final CIString asSetSource = updatedObject.getValueForAttribute(AttributeType.SOURCE);
+
+        if (flatAsSet){
+            final RpslObject asSetObject = rpslObjectDao.getByKeyOrNull(ObjectType.AS_SET, asSetKey);
+            if (asSetObject == null){
```

Comment:
Original object is saved in the updateContext, you do not need to hit the database?

Reply:
mmmm...I only see the object of the update request I don't see the one from database

Reply:
done